base/src/external.rs

```diff
@@ -0,0 +1,97 @@
+use std::{ffi::c_void, ops::Deref};
+
+use lean_sys::{
+    lean_alloc_external, lean_apply_1, lean_external_class, lean_get_external_class,
+    lean_get_external_data, lean_inc, lean_is_exclusive, lean_object, lean_register_external_class,
+};
+
+use crate::{Layout, Obj, TObj, TObjRef};
+
+pub trait AsExternalObj: Clone + 'static {
+    type ObjIter<'a>: Iterator<Item = &'a Obj>;
+    fn obj_iter(&self) -> Self::ObjIter<'_>;
+
+    unsafe extern "C" fn foreach(data: *mut c_void, closure: *mut lean_object) {
+        let data = data as *mut Self;
+        for i in (*data).obj_iter() {
+            lean_inc(i.0);
+            lean_inc(closure);
+            lean_apply_1(closure, i.0);
+        }
+    }
```

Comment:
I think it is better to leave this as an internal iteration, and have a (unsafe) trait for rust types to support traversal to find all the embedded lean objects.

Reply:
Yes that's right, except instead of "let the user implement this on their own" it's more like "have a procedural macro that does the implementing for them". That's future work though, for now the trait itself and some baseline implementations is sufficient.